### PR TITLE
introduce resolve-includes.groovy script

### PR DESCRIPTION
**Pull request by jstastny-cz**
Dynamic control over includes/excludes using groovy script.

### CODE REVIEW

**Review by baldimir — APPROVED** [review]

**Review by radtriste — APPROVED** [review]

File: pom.xml [hunk 1]
```diff
@@ -32,12 +32,15 @@
     <invoker.plugin.version>3.1.0</invoker.plugin.version>
     <unpack.build.version>1.0-SNAPSHOT</unpack.build.version>
     <dependency.plugin.version>3.1.2</dependency.plugin.version>
+    <gmavenplus.plugin.version>1.12.1</gmavenplus.plugin.version>
+    <groovy.version>2.5.14</groovy.version>
```
Comment (jomarko) [inline pom.xml:36]: do we want to try latest available 3.0.8?

File: pom.xml [hunk 2]
```diff
@@ -56,8 +59,32 @@
     <jbpm.wb.repo.url>${git.server.url}/jbpm-wb</jbpm.wb.repo.url>
     <kie.wb.common.repo.url>${git.server.url}/kie-wb-common</kie.wb.common.repo.url>
     <kogito.examples.repo.url>${git.server.url}/kogito-examples</kogito.examples.repo.url>
+
+    <!-- properties affecting maven modules resolution using resolve-includes.groovy script -->
+    <maven.modules.resolution.timeout>600000</maven.modules.resolution.timeout>
```
Comment (jomarko) [inline pom.xml:64]: is this miliseconds?

File: run-tests-with-build-kogito-runtimes/pom.xml [hunk 3]
```diff
@@ -48,96 +48,40 @@
...
-            <pomInclude>kogito-runtimes*/kogito-quarkus-parent/kogito-quarkus-rules-extension/kogito-quarkus-rules-integration-test/pom.xml</pomInclude>
-            <pomInclude>kogito-runtimes*/kogito-quarkus-parent/kogito-quarkus-rules-extension/kogito-quarkus-rules-integration-test-hot-reload/pom.xml</pomInclude>
-            <pomInclude>kogito-runtimes*/kogito-quarkus-parent/kogito-quarkus-rules-extension/kogito-quarkus-rules-integration-test-legacy/pom.xml</pomInclude>
-            <pomInclude>kogito-runtimes*/kogito-quarkus-parent/kogito-quarkus-rules-extension/kogito-quarkus-rules-integration-test-session/pom.xml</pomInclude>
-
-            <!-- kogito-workitems -->
-            <pomInclude>kogito-runtimes*/kogito-workitems/pom.xml</pomInclude>
+            <!-- includes everything, filtering is done in resolve-includes.groovy script
+            which adds invoker-run.groovy script containing 'return false' near each pom.xml
+            that should be excluded.-->
```
Comment (jomarko) [inline run-tests-with-build-kogito-runtimes/pom.xml:80]: The readme is now out of date for run-tests-with-build/kogito-runtimes submodule

File: scripts/resolve-includes.groovy [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+/**
+ * Recursively find all folders that contain pom.xml
+ * @return list of string paths
+ */
+private List<String> findAllDirsWithPom() {
+    File base = new File(basedir)
+    List<File> poms = new ArrayList<>()
+    base.traverse(type: groovy.io.FileType.FILES, nameFilter: "pom.xml") { it -> poms.add(it) }
+    return poms.stream().map({pom -> trimPrefix(pom.toPath().getParent().toAbsolutePath(), base.toPath())}).collect(Collectors.toList())
```
Comment (jomarko) [inline scripts/resolve-includes.groovy:91]: kiegroup formatting format this line fo me as:
```
    return poms.stream().map({ pom -> trimPrefix(pom.toPath().getParent().toAbsolutePath(), base.toPath()) }).collect(Collectors.toList())
```

File: scripts/resolve-includes.groovy [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    }
+    return mvnCommandToRun
+}
+
+/**
+ * Method to prevent execution of provided projects by adding invoker-specific script into project folder.
+ * @param toExclude list of projects to exclude
+ */
+private void excludeFromInvokerExecution(List toExclude) {
+    def ingoredLog = "ignored-interim-poms.log"
```
Comment (jomarko) [inline scripts/resolve-includes.groovy:130]: small typo

**Review by jomarko — COMMENTED** [review]
I was not able to run kogito-runtimes submodule locally.

Output I got
```
[INFO] Using plugin classloader, includes GMavenPlus and project classpath.
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.codehaus.groovy.reflection.CachedClass (file:/home/jomarko/.m2/repository/org/codehaus/groovy/groovy/2.5.14/groovy-2.5.14.jar) to method java.lang.Object.finalize()
WARNING: Please consider reporting this to the maintainers of org.codehaus.groovy.reflection.CachedClass
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
[INFO] Using Groovy 2.5.14 to perform execute.
[INFO] Running Groovy script from file:/home/jomarko/redhat/github/kiegroup/run-tests-with-build/scripts/resolve-includes.groovy.
[INFO] Resolving Maven reactor projects for command 'mvn -q -am exec:exec -Dexec.executable=pwd -Dproductized -Dmaven.repo.local=/home/jomarko/.m2/repository -s /home/jomarko/redhat/github/kiegroup/run-tests-with-build/.mvn/settings.xml'
[INFO] Waiting up to 600000 ms for resolution to end.
[INFO] Standard output will be available in /home/jomarko/redhat/github/kiegroup/run-tests-with-build/run-tests-with-build-kogito-runtimes/target/mvn-std-out.log after the build finishes.
[ERROR] Maven build unsuccessful with code 1, see /home/jomarko/redhat/github/kiegroup/run-tests-with-build/run-tests-with-build-kogito-runtimes/target/mvn-std-err.log.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  18.400 s
[INFO] Finished at: 2021-05-06T14:37:52+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.codehaus.gmavenplus:gmavenplus-plugin:1.12.1:execute (resolve-includes) on project run-tests-with-build-kogito-runtimes: Error occurred while calling a method on a Groovy class from classpath.: InvocationTargetException: Mvn command ended with error code 1 -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

The referenced log file is empty
```
/home/jomarko/redhat/github/kiegroup/run-tests-with-build/run-tests-with-build-kogito-runtimes/target/mvn-std-err.log
```

I started as:
```
mvn clean install -Dkogito.download.build.version=1.5.0.redhat-00001 -f run-tests-with-build-kogito-runtimes/
```

As last thing, I miss somewhere the explanation we discussed privately, why needed to run just leaves modules for the modules, where projhect-sources.tar.gz is used as input.

File: pom.xml (hunk 1 above)
Comment (jstastny-cz) [reply pom.xml:36]: Should be possible, http://groovy.github.io/GMavenPlus/dependencies.html states 3.0.7 as test dependency.

File: pom.xml (hunk 2 above)
Comment (jstastny-cz) [reply pom.xml:64]: yes, will rename the prop to indicate that.

File: scripts/resolve-includes.groovy (hunk 4 above)
Comment (jstastny-cz) [reply scripts/resolve-includes.groovy:91]: My bad, will reformat using kiegroup template.

File: scripts/resolve-includes.groovy (hunk 5 above)
Comment (jstastny-cz) [reply scripts/resolve-includes.groovy:130]: Will fix.

File: run-tests-with-build-kogito-runtimes/pom.xml (hunk 3 above)
Comment (jstastny-cz) [reply run-tests-with-build-kogito-runtimes/pom.xml:80]: I've corrected that statement and added to top-level readme details about the approach and implementation. Please have a look, committing in a bit.

**Review by jomarko — APPROVED** [review]
Thank you, now tests were started without an issue for:
- kogito-runtimes
- kogito-examples

**Pull request merged**